compiler/rustc_const_eval/src/interpret/step.rs

```diff
@@ -90,6 +90,14 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                 self.write_discriminant(*variant_index, &dest)?;
             }
 
+            Finalize(place) => {
+                if M::enforce_validity(self) {
+                    let place = self.eval_place(**place)?;
+                    // Invariant: when `Finalize` is invoked on a place, that place is fully valid
+                    self.validate_operand(&self.place_to_op(&place)?)?;
+                }
+            }
```

Comment:
🤦 this is unreachable, I'm erasing all `Finalize` before going into `optimized_mir`, because it's not useful after that

Reply:
`RemoveFinalize` looks like it's currently gated on `mir_opt_level > 0`

Reply:
Ah, I remember why I added it. The const propagator runs on MIR that still has `Finalize` statements

Reply:
Should Miri run this, or not?

`enforce_validity` is anyway false for CTFE.

Reply:
>This does not look like it reports a mutation/store for the place

I think this is indeed a read-only (but potentially UB) operation. Why would it report a store?

Reply:
Ah, looks like the semantics changed.

I agree this should be read-only. However I think it makes sense to actually give it semantics (i.e., potential UB) -- that justifies its role in the analyses much better.

Reply:
> Should Miri run this, or not?
> 
> `enforce_validity` is anyway false for CTFE.

Const prop uses validity checking to ensure it doesn't propagate half baked values

Reply:
Oh, interesting. Seems like an expensive way to do that but sure.^^

Reply:
> Const prop uses validity checking to ensure it doesn't propagate half baked values

Conceptually this is very strange though. Optimizations should be allowed to rely on the fact that values are valid, they should never have to run validity to check whether they are valid.

Reply:
Oh, we do this for the result of const prop. As we use undef as a marker for "unknown" we can end up with aggregates that are partially undef. We don't want to write those back to MIR, because those undefs usually mean "a value only known at runtime"

Reply:
I think that's just unsound? If you have a variable of type `MaybeUninit` that is partially initialized but the value it is initialized to is unknown, then validation will be like "sure uninit is fine here" but actually the optimization is wrong.

There is a big difference between "this memory is definitely uninit" and "we don't know the contents of this memory", and it sounds like const prop is conflating these two things?

Reply:
Yea, we knew this wasn't going to scale to some types. We'd need a different kind of marker for unknown data, but I also don't know if we can do that without complicating the interpreter. Maybe const prop should be its own thing... but it's also sad if we duplicate work.

Reply:
const prop does symbolic evaluation. It is not surprising that a concrete interpreter is not up for the task.

It might be possible to generalize parts of the interpreter to support symbolic and concrete evaluation, but we probably don't want that for the full thing. (const prop doesn't hit all parts of the interpreter anyway.)

Currently, what do we do to prevent soundness issues like the one I described?

Reply:
We avoid propagating most aggregates.

I think the right level of code sharing would be to call the various math operations manually and not reuse any of the functions that evaluate a mir statement or rvalue

Reply:
Yeah, I agree. Basically `operand.rs` and `cast.rs`.